### DISCUSSION TITLE
Whats the 'lazy' way to write the `init` part for a plugin?

### DISCUSSION THREAD

**Discussion by davidgao7** [LazyVim/LazyVim#2966] [Q&A] [answered]
Greeting Lazyvim community,

Does anyone happen to know how to "import" a plugin and call the init/setup function in lazyvim?
I was trying to adapt git-blame.nvim by f-person to lazyvim

```lua
-- TODO: how to setup the init
return {
    {
      -- git blame plugin
      "f-person/git-blame.nvim",
      init = function()
        -- git blame template
        -- Available options: <author>, <committer>, <date>, <committer-date>, <summary>, <sha>
        vim.g.gitblame_message_template = " <summary> • <date> • <author>"
        -- enables git-blame.nvim on Neovim startup
        vim.g.gitblame_enabled = 1
        -- start virtual text at column
        -- Have the blame message start at a given column instead of EOL.
        -- If the current line is longer than the specified column value
        -- the blame message will default to being displayed at EOL.
        vim.g.gitblame_virtual_text_column = 1
      end,

      keys = {
        {
          "<leader>gbu",
          "<cmd>GitBlameToggle<cr>",
          desc = "toggle git blame",
        },
        {
          "<leader>gbe",
          "<cmd>GitBlameEnable<cr>",
          desc = "enable git blame",
        },
        {
          "<leader>gbd",
          "<cmd>GitBlameDisable<cr>",
          desc = "disable git blame",
        },
        {
          "<leader>gbh",
          "<cmd>GitBlameCopySHA<cr>",
          desc = "copy line commit SHA",
        },
        {
          "<leader>gbl",
          "<cmd>GitBlameCopyCommitURL<cr>",
          desc = "copy line commit URL",
        },
        {
          "<leader>gbo",
          "<cmd>GitBlameOpenFileURL<cr>",
          desc = "opens file in default browser",
        },
        {
          "<leader>gbc",
          "<cmd>GitBlameCopyFileURL<cr>",
          desc = "copy file url to clipboard",
        },
      },
    },
    {
      "folke/which-key.nvim",
      opts = {
        defaults = {
          ["<leader>gb"] = { name = "git blame+" },
        },
      },
    },
}
```

**Comment by dpetka2001** [comment]
The way you did it is fine. I have it like that and works fine.

**Reply by davidgao7** [reply]
@dpetka2001 sorry for the late reply, 

Is there a way to set the plugin to turn on whenever opening a git buffer? 

I find that If I just let lazy import repo like this
```lua
return{
    "f-person/git-blame.nvim"
}
```
I could see the git blames right after opening a git buffer. without do anything further.

once I add more settings, that's where the problem starts to show.

[I was asking suggestions from f-person too](https://github.com/f-person/git-blame.nvim/issues/129#issuecomment-2045177731), I'm not sure how lazy.nvim/lazyvim config the plugin. Perhaps the init/setup function for git-blame.nvim hasn't been called yet right after I'm in neovim?

**Reply by dpetka2001** [reply]
I just have a simple config for it and it works out of the box. I don't do anything particular. I just load the plugin whenever I'm in a git repo, that's all.

I believe in your case it's because you're using `keys` to lazy-load the plugin. The option `vim.g.gitblame_enabled = 1` won't take effect, since `lazy.nvim` will use the `keys` handlers to lazy-load the plugin either way. I don't really understand what you want to achieve.

In the other issue you're saying
> When I open a git file I have to manually call the command < GitBlameEnable> by using my shortcut to see the git blame,

That's just how `lazy.nvim` works. If you don't want that, then just don't use `keys` and instead configure your keymaps either in `opts` or `config` or even `keymaps.lua`.

**Comment by davidgao7** [comment, marked as answer]
I think I've figured it out! For people who are still struggling the same problem, check out this [issue](https://github.com/f-person/git-blame.nvim/issues/129#issuecomment-2051818749)
